contrib/generic_proxy/filters/network/source/router/router.cc

```diff
@@ -280,17 +343,38 @@ void RouterFilter::kickOffNewUpstreamRequest() {
     return;
   }
 
-  auto tcp_data = thread_local_cluster->tcpConnPool(Upstream::ResourcePriority::Default, this);
-  if (!tcp_data.has_value()) {
+  if (callbacks_->boundUpstreamConn().has_value()) {
+    // Upstream connection binding is enabled and the upstream connection is already bound.
+    // Create a new upstream request without a connection pool and start the request.
+    auto upstream_request = std::make_unique<UpstreamRequest>(*this, absl::nullopt);
+    auto raw_upstream_request = upstream_request.get();
+    LinkedList::moveIntoList(std::move(upstream_request), upstream_requests_);
+    raw_upstream_request->startStream();
+    return;
+  }
+
+  auto pool_data = thread_local_cluster->tcpConnPool(Upstream::ResourcePriority::Default, this);
+  if (!pool_data.has_value()) {
     filter_complete_ = true;
     callbacks_->sendLocalReply(Status(StatusCode::kUnavailable, "no_healthy_upstream"));
     return;
   }
 
-  auto upstream_request = std::make_unique<UpstreamRequest>(*this, std::move(tcp_data.value()));
+  if (protocol_options_.bindUpstreamConnection()) {
+    // Upstream connection binding is enabled and the upstream connection is not bound yet.
+    // Bind the upstream connection and start the request.
+    callbacks_->bindUpstreamConn(std::move(pool_data.value()));
+    auto upstream_request = std::make_unique<UpstreamRequest>(*this, absl::nullopt);
+    auto raw_upstream_request = upstream_request.get();
+    LinkedList::moveIntoList(std::move(upstream_request), upstream_requests_);
+    raw_upstream_request->startStream();
+    return;
+  }
+
+  // Normal upstream request.
+  auto upstream_request = std::make_unique<UpstreamRequest>(*this, std::move(pool_data.value()));
```

Comment:
Can we use the filter's UpstreamManager for the case of `bindUpstreamConnection == false`? Like, we can reset the filter's UpstreamManager when the request is done, then it will create new UpstreamManager when the new request is coming.

I thought if that works, that will simple the code. The different of bind upstream connection or not is just release the Filter's UpstreamManager or not when the request is done.

Reply:
It‘s possbile to accept multiple downstream request at same time. So, if `bindUpstreamConnection == false`, it would be better to manage the upstream connection by the upstream request self. Especially considering that the coming retry support.